lib/Parse/ParsePattern.cpp

```diff
@@ -648,6 +648,26 @@ Parser::parseFunctionSignature(Identifier SimpleName,
   }
 
   SourceLoc arrowLoc;
+
+  auto diagnoseInvalidThrows = [&]() -> Optional<InFlightDiagnostic> {
+    if (throwsLoc.isValid())
+      return None;
+
+    if (Tok.is(tok::kw_throws)) {
+      throwsLoc = consumeToken();
+    } else if (Tok.is(tok::kw_rethrows)) {
+      throwsLoc = consumeToken();
+      rethrows = true;
+    }
+
+    if (throwsLoc.isValid()) {
+      auto diag = rethrows ? diag::rethrows_after_function_result
+                           : diag::throws_after_function_result;
```

Comment:
The names of these diagnostics are a little misleading now. `throws_in_incorrect_place`?